erc20/src/lib.rs

```diff
@@ -2,12 +2,15 @@
 #![no_main]
 
 use core::default::Default;
-
 use contract_derive::contract;
-use eth_riscv_runtime::{types::Mapping, log};
-
+use eth_riscv_runtime::{types::Mapping};
 use alloy_core::primitives::{Address, address};
+use alloy_sol_types::sol;
 
+sol! {
```

Comment:
hmmm I'm not sure we should be using `sol!` here, ideally we never force the user to use such macros.
@gakonst opinions?

I think for this PR we should simply do `emit("Transfer", fro, to, value)`, where `emit` is a library function.
Then in the future we can add rules to contract derive to do that automatically for functions annotated witt `#[event]` or something like that.

Reply:
I hada similar thought, yes I think so as well, keeping the user away from having to know about Solidity-style syntax. 

I would let the PR get merged in the interest of getting the feature in and iterate if we think that's harder to do though

Reply:
I just wrote a quick solution to do it as you suggested. This would only be the quick option; later, we could use #[event] attributes to make it much more customizable and easier to use.

For now, I’m sharing this to get your opinion:
@leonardoalt 
@gakonst 

In contract-derive/src/lib.rs:
```solidity
    let emit_helper = quote! {
        fn emit(event_name: &str, from: Address, to: Address, value: u64) {
            use alloy_sol_types::SolValue;
            use alloy_core::primitives::keccak256;
            
            let signature = [event_name.as_bytes(), b"(address,address,uint64)"].concat();
            let topic0 = keccak256(&signature);
            
            let mut all_topics = [0u8; 96]; 
            all_topics[0..32].copy_from_slice(topic0.as_slice());
            
            let from_encoded = from.abi_encode();
            let to_encoded = to.abi_encode();
            
            all_topics[32..64].copy_from_slice(&from_encoded);
            all_topics[64..96].copy_from_slice(&to_encoded);
            
            let value_data = value.abi_encode();
            
            log(
                value_data.as_ptr() as u64,
                value_data.len() as u64,
                all_topics.as_ptr() as u64,
                3 // topics
            );
        }
    };



    let call_method = quote! {
      ...

        
        #emit_helper

        impl Contract for #struct_name {
```

contract:

```solidity
#![no_std]
#![no_main]

use core::default::Default;
use contract_derive::contract;
use eth_riscv_runtime::{types::Mapping};
use alloy_core::primitives::{Address, address};
use alloy_sol_types::sol;


#[derive(Default)]
pub struct ERC20 {
    balance: Mapping<Address, u64>,
}

#[contract]
impl ERC20 {
    pub fn balance_of(&self, owner: Address) -> u64 {
        self.balance.read(owner)
    }

    pub fn transfer(&self, from: Address, to: Address, value: u64) -> bool {
        let from_balance = self.balance.read(from);
        let to_balance = self.balance.read(to);

        if from == to || from_balance < value {
            revert();
        }

        self.balance.write(from, from_balance - value);
        self.balance.write(to, to_balance + value);

        emit("Transfer", from, to, value);
        
        true
    }
```